### PR TITLE
Register wf obligation before normalizing in wfcheck 

### LINKED ISSUE #100041: Normalization can skip WF
I tried this code: [[playground]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c5cd930a4dee00e88a93712e13ee57e0)

```rust
trait WellUnformed {
    type RequestNormalize;
}

impl<T: ?Sized> WellUnformed for T {
    type RequestNormalize = ();
}

const _: <[[[[[[u8]]]]]] as WellUnformed>::RequestNormalize = ();
const _: <Vec<str> as WellUnformed>::RequestNormalize = ();
```

I expected this to error; `[[[[[[u8]]]]]]` is clearly not a well-formed type, as
> error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
> = help: the trait `Sized` is not implemented for `[u8]`
> = note: slice and array elements must have `Sized` type

`Vec<str>` is perhaps even more cursed.

Instead: it compiles.

This is especially relevant for blanket/builtin traits which have associated types, such as `mem::Discriminant` and `ptr::Pointee`.

### Meta

1.64.0-nightly (2022-07-31 f9cba63746d0fff81625)

@rustbot label +A-associated-items

cc @eddyb who said this was
@rustbot label +I-unsound

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #100041.

Why are we normalizing in wfcheck at all? Should we just be dealing w unnormalized types? I guess whether we normalize or not at all has effects on implied bounds...

cc @rust-lang/types

r? types

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 1]
```diff
@@ -288,7 +288,13 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
 
         let infcx = self.selcx.infcx();
 
-        if obligation.predicate.has_projections() {
+        // Normalizing WellFormed predicates is not sound, see #100041
+        if obligation.predicate.has_projections()
+            && !matches!(
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:293]: This shouldn't be necessary once @lcnr's PR #99217 lands, and we should use `obligation.predicate.allow_normalization()` here instead.

**Comment by lcnr** [comment]
this is a breaking change again, for similar reasons to #99217.

I think we would want to merge these two PRs and get another crater run in :sweat_smile: 

apart from that lgtm
r? @lcnr

**Comment by lcnr** [comment]
> Why are we normalizing in wfcheck at all?

anything that's not wf checking pretty much expects normalized results :sweat_smile: without lazy norm having unnormalized projections tends to be bad :tm: we do normalize in `fulfill` so for obligations that requirement doesn't really exist anymore[^1], but e.g. `combine` (`infcx.at(cause, param_env).eq(a, b)`) still expect `a` and `b` to be normalized as much as possible.

[^1]: maybe we do want to completely stop normalizing obligations before adding them as that should improve error messages, not sure about the perf impact.

**Comment by compiler-errors** [comment]
@lcnr yeah, I'm more curious about whether it's correct/important to have normalized types before we add them to the implied bounds list, like here: https://github.com/rust-lang/rust/blob/master/compiler/rustc_typeck/src/check/wfcheck.rs#L1523

This PR doesn't affect that, for example (just moves the WF obligation on that fn sig's types to before they get normalized). Should we be normalizing types there at all?

**Comment by lcnr** [comment]
> @lcnr yeah, I'm more curious about whether it's correct/important to have normalized types before we add them to the implied bounds list, like here: https://github.com/rust-lang/rust/blob/master/compiler/rustc_typeck/src/check/wfcheck.rs#L1523
> 
> This PR doesn't affect that, for example (just moves the WF obligation on that fn sig's types to before they get normalized). Should we be normalizing types there at all?

With the changes to normalization in fulfill, and maybe some other ones of https://github.com/rust-lang/rust/pull/99217, we are allowed to (and should) add both. We do need to normalize stuff before we currently don't elaborate implied bounds so if you have `<T as Trait<'a>>::Assoc` which results in `&'a T` just adding `<T as Trait<'a>>::Assoc` as an implied bound doesn't allow you to assume `T: 'a` inside of the function. If we merge your PR with #99217 we should check that every time we add implied bounds, we add both normalized and unnormalized ones (alternatively, add unnormalized ones and normalize in `add_implied_bounds` :grin:)

**Comment by compiler-errors** [comment]
Alright. That makes sense, thanks for the explanation.

As a first step I can cherry-pick #99217 onto this branch so we can do another crater run of the merged code.

I'll think a bit more about the best way that we can collect both the normalized and un-normalized forms of those types in wfcheck.

**Comment by compiler-errors** [comment]
@craterbot check

**Comment by compiler-errors** [comment]
> We do need to add the normalized ones as we currently don't elaborate implied bounds

ok crater shows clearly that yeah i need to do this then

**Comment by compiler-errors** [comment]
Ok, so now we record both the normalized and unnormalized tys, which should probably fix the regressions. I will check on crater once this try-build is done.

**Comment by compiler-errors** [comment]
@craterbot check

(i wonder if we just need to check this with the list of failed crates...)

**Comment by lcnr** [comment]
yeah, i think we only want to rerun, might make sense to increase the priority for this

@craterbot abort

**Comment by lcnr** [comment]
@craterbot run crates=https://crater-reports.s3.amazonaws.com/pr-100046/retry-regressed-list.txt

**Comment by lcnr** [comment]
@craterbot abort

**Comment by lcnr** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-100046/retry-regressed-list.txt p=1

sorry for the spam :sweat_smile:

**Comment by compiler-errors** [comment]
for the record, passing `list=` and `p=` can be done w/o aborting the experiment i think

**Comment by compiler-errors** [comment]
There only seems to be one legitimate new failure due to this PR, and it actually seems correct to be failing, since it's an unnormalized type that required `T: 'a` (and that requirement probably goes away when you normalize?)

Though I will defer to @lcnr to review more closely.

**Comment by lcnr** [comment]
yes, the only regression found seems to be https://github.com/itsybitesyspider/retriever

more concretely https://github.com/itsybitesyspider/retriever/blob/7fc9d9a97f1d9b8c55f490d377cc88b24f2c9de5/src/queries/secondary_index.rs#L488-L499:
```rust
impl<'a, K, T, S> KeySet<'a, K> for HashSet<T, S>
where
    T: Clone + Borrow<K> + 'a,
    K: ToOwned<Owned = T> + ?Sized + 'a,
{
    #[allow(clippy::type_complexity)]
    type KeySetIter = Map<<&'a HashSet<T, S> as IntoIterator>::IntoIter, fn(&'a T) -> Cow<'a, K>>;

    fn iter_keys(&'a self) -> Self::KeySetIter {
        self.iter().map(|t: &'a T| Cow::Borrowed(t.borrow()))
    }
}
```
The issue here is that `&'a HashSet<T, S>` requires `S: 'a` but because `<&'a HashSet<T, S> as IntoIterator>::IntoIter` normalizes to `Iter<'a, T>` this bound gets removed by normalization.

-----

I believe that checking wf before normalization is correct and the current state should be considered a bug. As the fallout seems to be manageable I propose we merge this PR as is, while opening a PR to https://github.com/itsybitesyspider/retriever/ fixing this issue by either using the normalized type directly or by adding an `S: 'a` bound to the impl.

Any regressions caused by this PR can be fixed by directly using the associated type instead of the projection.

@rfcbot merge

While I believe this to pretty much exclusively be the responsibility of t-types, I am not fully confident about that, so cc @rust-lang/lang to make sure that we're not overstepping our bounds here.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed 

Agreed we should be checking WF before normalization.

**Comment by compiler-errors** [comment]
Lol, so this PR cannot land as-is anymore because of #100676 and a few other PRs. I need to rework this now.

**Comment by compiler-errors** [comment]
@lcnr I think this is blocked on your work to get both the unnormalized and normalized versions of a type to be considered implied wf types, correct? Since e6b8a2129fdfe4fcde1e418a8b3ec479a23c87f1 is obsolete (we don't collect implied wf tys here anymore).

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 2]
```diff
@@ -288,7 +288,8 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
 
         let infcx = self.selcx.infcx();
 
-        if obligation.predicate.has_projections() {
+        // Normalizing WellFormed predicates is not sound, see #100041
+        if obligation.predicate.has_projections() && obligation.predicate.allow_normalization() {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:292]: already dealt with by `try_normalize_with_depth_to`, no need to check for `allow_normalization` here

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/fulfill.rs:292]: Oh cool

**Comment by lcnr** [comment]
blocked on implied bounds for unnormalized types, should hopefully implement this over the next few weeks

**Comment by aliemjay** [comment]
#105948 should probably be fixed first. Otherwise this will break the following arguably correct code:
```rust
trait Trait { type Ty; }
impl<T> Trait for T { type Ty = (); }
fn pass<T>(_: <&'static T as Trait>::Ty) {}
```

**Comment by lcnr** [comment]
closing this PR as https://github.com/rust-lang/rust/pull/100046#issuecomment-1243953943 is blocked on `-Ztrait-solver=next` plus additional work, so that's going to take a while

**Comment by lcnr** [comment]
we need the unnormalized implied bounds for impl headers as well. Fixing #105948 would only fix it in function signatures :thinking: I think the following test breaks with this PR:
```rust
trait Trait {}

trait ToUnit {
    type Assoc;
}
impl<T> ToUnit for T {
    type Assoc = ();
}

impl<'a, 'b> Trait for <&'a &'b () as ToUnit>::Assoc {}
```

**Pull request closed without merging**